File: xmetrics_file.R
#!/usr/bin/env Rscript

source("utils.r")

args <- commandArgs(trailingOnly = TRUE)
data_file <- if (length(args) >= 1L) args[[1L]] else "xmetrics_data.txt"
n_bkps <- 3L
min_size <- 40L
margin <- 40L

t0 <- proc.time()[["elapsed"]]
signal <- read_series(data_file)
true_bkps <- read_true_bkps(data_file)
n <- length(signal)
width <- max(80L, n %/% 10L)

cat("file =", data_file, "\n")
cat("n =", n, "\n")
cat("true bkps =", paste(true_bkps, collapse = " "), "\n")
cat("n_bkps =", n_bkps, "\n")
cat("min_size =", min_size, "\n")
cat("margin =", margin, "\n")
cat("\n")

methods <- list(
  list(name = "binseg", run = function() binseg_l2_1d(signal, n_bkps, min_size)),
  list(name = "bottomup", run = function() bottomup_l2_1d(signal, n_bkps, min_size, jump = 5L)),
  list(name = "window", run = function() window_l2_1d(signal, n_bkps, width, jump = 5L)),
  list(name = "dynp", run = function() solve_dynp_cost_1d(signal, n_bkps, min_size, "l2"))
)

for (item in methods) {
  t1 <- proc.time()[["elapsed"]]
  bkps <- item$run()
  elapsed <- proc.time()[["elapsed"]] - t1
  pr <- precision_recall_cpt(true_bkps, bkps, margin)
  hd <- hausdorff_cpt(true_bkps, bkps)
  ri <- randindex_cpt(true_bkps, bkps)

  cat("method =", item$name, "\n")
  cat("estimated bkps =", paste(bkps, collapse = " "), "\n")
  cat(sprintf("precision = %.3f\n", pr[["precision"]]))
  cat(sprintf("recall = %.3f\n", pr[["recall"]]))
  cat(sprintf("hausdorff = %.3f\n", hd))
  cat(sprintf("randindex = %.6f\n", ri))
  cat(sprintf("elapsed seconds = %.3f\n", elapsed))
  cat("\n")
}

cat(sprintf("wall time elapsed (s) = %.3f\n", proc.time()[["elapsed"]] - t0))

File: xmetrics_file.py
"""
Read a univariate series from a text file and evaluate changepoint estimates
with ruptures metrics.
"""

from __future__ import annotations

import sys
import time

import numpy as np
import ruptures as rpt
from ruptures.metrics import hausdorff, precision_recall, randindex


def read_true_bkps(path: str) -> list[int]:
    with open(path, encoding="ascii") as f:
        for line in f:
            if line.startswith("# true_bkps ="):
                return [int(x) for x in line.split("=", 1)[1].split()]
    raise ValueError("input file is missing '# true_bkps =' metadata")


def main() -> None:
    t0 = time.perf_counter()
    data_file = sys.argv[1] if len(sys.argv) > 1 else "xmetrics_data.txt"
    n_bkps = 3
    min_size = 40
    margin = 40

    signal = np.loadtxt(data_file, comments="#", ndmin=1)
    if signal.ndim != 1:
        signal = np.ravel(signal)
    true_bkps = read_true_bkps(data_file)
    n = signal.shape[0]

    methods = [
        ("binseg", rpt.Binseg(model="l2", min_size=min_size, jump=1)),
        ("bottomup", rpt.BottomUp(model="l2", min_size=min_size)),
        ("window", rpt.Window(width=max(80, n // 10), model="l2")),
        ("dynp", rpt.Dynp(model="l2", min_size=min_size, jump=1)),
    ]

    print("file =", data_file)
    print("n =", n)
    print("true bkps =", true_bkps)
    print("n_bkps =", n_bkps)
    print("min_size =", min_size)
    print("margin =", margin)
    print()

    for name, algo in methods:
        t1 = time.perf_counter()
        bkps = algo.fit(signal).predict(n_bkps=n_bkps)
        elapsed = time.perf_counter() - t1
        precision, recall = precision_recall(true_bkps, bkps, margin=margin)
        hd = hausdorff(true_bkps, bkps)
        ri = randindex(true_bkps, bkps)

        print("method =", name)
        print("estimated bkps =", bkps)
        print(f"precision = {precision:.3f}")
        print(f"recall = {recall:.3f}")
        print(f"hausdorff = {hd:.3f}")
        print(f"randindex = {ri:.6f}")
        print(f"elapsed seconds = {elapsed:.3f}")
        print()

    print(f"wall time elapsed (s) = {time.perf_counter() - t0:.3f}")


if __name__ == "__main__":
    main()
